Create endpoint for schedule generation

glasierr opened the pull request:
resolves #94

Commit:
Create endpoint for schedule generation

glasierr commented:
Payload for the request:

{
	"description": "Cool schedule",
    "availability": {
        "weekDays": ["TUESDAY", "MONDAY", "WEDNESDAY"],
        "startDate": "2017-03-02",
        "endDate": "2017-04-03",
        "exceptionDays": []
    },
	"startTimeInbound": "13:45",
    "endTimeInbound": "17:45",
    "startTimeOutbound": "13:45",
    "endTimeOutbound": "18:45",
    "headway": 2100,
    "dwellTime": 50,
    "averageSpeed": "10 m/s",
    "minLayover": 600,
    "inboundRouteId": 1,
    "outboundRouteId": 2
}

wingsofovnia commented:
What is the case for separating startDate and startTimeInbound? What is it not startTimeInbound = new Date() ?

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -0,0 +1,68 @@
+import static org.springframework.web.bind.annotation.RequestMethod.POST;
+
+@RestController
+@RequestMapping(path = "/schedules")
```
lets highlight it somehow that it's a generator, not a regular schedule resource. e.g. `/schedules/generators`

glasierr replied:
Agree with highlighting, but I think that /generators will confuse, because we are not working with some generator resources, so I prefer to have it as noun, e.g. 'generate'. It points that we are working with schedules and doing a generation of it

wingsofovnia replied:
generate is a verb :)

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -0,0 +1,68 @@
+    @Autowired
+    private RouteRepository routeRepository;
+
+    @RequestMapping(method = POST, path = "/generate")
```
and `/basic` here. `public Schedule generateBasicSchedule(..)`

wingsofovnia replied:
btw are you sure it should be POST? seems it doesn't save a generated Schedule so it's more GET, not POST

glasierr replied:
Agree with using GET instead of POST, however it also does not seem to 100% fit

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/exception/NotFoundException.java:
```diff
@@ -0,0 +1,20 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.NOT_FOUND)
+public class NotFoundException extends RuntimeException {
```
Imho it's not necessary to create yet another exception. You can just return Resource.notFound():
https://github.com/socialedge/vega/blob/develop/backend-account/src/main/java/eu/socialedge/vega/backend/account/application/api/operator/OperatorController.java#L62

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/exception/ScheduleGeneratorException.java:
```diff
@@ -1,4 +1,4 @@
-package eu.socialedge.hermes.backend.schedule.domain;
+package eu.socialedge.hermes.backend.schedule.domain.exception;
```
Please keep it near the class (in the same package) it's supposed to be used.

glasierr replied:
It would be done, when yet another NotFound exception will be removed

wingsofovnia commented on backend-application/build.gradle:
```diff
@@ -11,4 +11,6 @@ dependencies {
     compile libraries.persistence.hibernateCore
     compile libraries.persistence.h2Driver
+
+    compile libraries.dataformat.jacksonDatatype
```
please indicate that it's not a generic "datatype" but java8time. e.g. `jacksonDatatypeJsr310`

Commit:
Remove yet another NotFoundException

Use GET method instead of POST in schedule generation endpoint
Rename dependency name in dradle file

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -31,23 +31,30 @@
 
 @RestController
-@RequestMapping(path = "/schedules")
+@RequestMapping(path = "/schedules/generate")
```
use plural resources e.g. generated

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -31,23 +31,30 @@
-    public Schedule generateSchedule(@RequestBody @NotNull @Valid ScheduleSpecification spec) {
-        val inboundRoute = findRoute(spec.inboundRouteId());
-        val outboundRoute = findRoute(spec.outboundRouteId());
+    @RequestMapping(method = GET)
```
?generator=basic

glasierr replied:
What about /schedules/basic ?

glasierr committed:
Make schedule creation endpoint save the schedule and return it's id

glasierr committed:
Differentiate schedule generation endpoints by header

glasierr commented:
Now request requires header "generator" with value "basic" in order to generate schedule with basic generator

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -0,0 +1,86 @@
+        val generatedSchedule = scheduleBuilder.generate();
+        val persistedSchedule = scheduleRepository.save(generatedSchedule);
+
+        val scheduleUri = UriComponentsBuilder.fromPath("/schedules/").path(persistedSchedule.id().toString()).build().toUri();
```
return full url please

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/repository/ScheduleRepository.java:
```diff
@@ -17,9 +17,14 @@
 public interface ScheduleRepository extends CrudRepository<Schedule, Long> {
+
+    @Override
+    @RestResource(exported = false)
```
hide update methods too

glasierr replied:
'save' works for both create and update, so update is already disabled

wingsofovnia commented on dependencies.gradle:
```diff
@@ -22,6 +22,9 @@ ext {
             lombok: '1.16.+'
         ],
+        dataformat: [
+            jacksonDatatypeJsr310: '2.8.7'
```
my proposition is to move it to commons or `extensions:` (`ext:`)
```
extensions: [
jacksonDatatypeJsr310...
```

wingsofovnia commented on backend-schedule/src/main/java/eu/socialedge/hermes/backend/schedule/domain/api/ScheduleGenerationApi.java:
```diff
@@ -0,0 +1,86 @@
+    @Autowired
+    private ScheduleRepository scheduleRepository;
+
+    @RequestMapping(method = POST, headers = BASIC_GENERATOR)
```
Add default method which will handle requests without header

glasierr committed:
Create default schedule generation endpoint
